### PR TITLE
zephyr: Add CONFIG_MCUBOOT_CLEANUP_RAM

### PR DESCRIPTION
**Pull request by de-nordic**
Add Kconfig option to cleanup RAM in MCUboot before passing control to an application.

### CODE REVIEW

File: boot/zephyr/main.c [hunk 1]
```diff
@@ -236,7 +236,34 @@ static void do_boot(struct boot_rsp *rsp)
     __set_CONTROL(0x00); /* application will configures core on its own */
     __ISB();
 #endif
+#if CONFIG_MCUBOOT_CLEANUP_RAM
+    __asm__ volatile (
+        /* rt->reset -> r0 */
```
Comment (nvlsianpu) [inline boot/zephyr/main.c:241]:
```suggestion
        /* vt->reset -> r0 */
```

File: boot/zephyr/main.c [hunk 2]
```diff
@@ -236,7 +236,34 @@ static void do_boot(struct boot_rsp *rsp)
...
+        "   sub     r2, r2, %4\n"
+        "   cbz     r2, out\n"
+        "   b       clear\n"
+        "out:\n"
+        "   dsb\n"
+        /* jump to reset vector of an app */
+        "   bx      r0\n"
+        :
+        : "r" (vt->reset), "i" (CONFIG_SRAM_BASE_ADDRESS),
+          "i" (CONFIG_SRAM_SIZE * 1024), "i" (0), "i" (sizeof(uint32_t))
```
Comment (nvlsianpu) [inline boot/zephyr/main.c:261]: ` "i" (sizeof(uint32_t))` operand is not used, assembly above used immediate value - which is just perfect.

**Review by nvlsianpu — CHANGES_REQUESTED** [review]

**Review by nvlsianpu — APPROVED** [review]

File: boot/zephyr/main.c (hunk 1 above)
Comment (de-nordic) [reply boot/zephyr/main.c]: Fixed

File: boot/zephyr/main.c (hunk 2 above)
Comment (de-nordic) [reply boot/zephyr/main.c]: Yeah, so I have hardcoded this to #4 and remove the parameter.

**Review by nordicjm — CHANGES_REQUESTED** [review]
Does not work with ARMv6:
```
/tmp/ccmsO9XM.s: Assembler messages:
/tmp/ccmsO9XM.s:495: Error: selected processor does not support `cbz r2,out' in Thumb mode
/tmp/ccmsO9XM.s:488: Error: invalid immediate: 536870912 is out of range
/tmp/ccmsO9XM.s:488: Error: value of 20000000 too large for field of 2 bytes at 000000c2
/tmp/ccmsO9XM.s:489: Error: invalid immediate: 32768 is out of range
```

**Review by nordicjm — COMMENTED** [review]
Bit confused here, different symbol has been changed and depends on m7, we don't have any cortex m7's

**Comment by de-nordic** [comment]
> Bit confused here, different symbol has been changed and depends on m7, we don't have any cortex m7's

I am now confused too, because it compiled and worked and the code is actually picked up.

**Comment by de-nordic** [comment]
> > Bit confused here, different symbol has been changed and depends on m7, we don't have any cortex m7's
> 
> I am now confused too, because it compiled and worked and the code is actually picked up.

Yeah, sure, because I did the change to wrong Kconfig.

File: boot/zephyr/Kconfig [hunk 3]
```diff
@@ -219,6 +219,12 @@ config MCUBOOT_CLEANUP_ARM_CORE
 	  start-up code which can cause a module fault and potentially make the
 	  module irrecoverable.
 
+config MCUBOOT_CLEANUP_RAM
+	bool "Perform RAM cleanup"
+	depends on CPU_CORTEX_M4
```
Comment (nordicjm) [inline boot/zephyr/Kconfig:224]: `|| CPU_CORTEX_M33`?

**Review by nordicjm — APPROVED** [review]

**Pull request merged**